### ISSUE TITLE
Issues redirecting programmatically with push()

### ISSUE THREAD

**Issue by micahblu** [ReactTraining/react-router#4950]
My setup:
```javascript
import { HashRouter } from 'react-router-dom'
import createHistory from 'history/createBrowserHistory'
import { Route } from 'react-router'

import { ConnectedRouter, routerReducer, routerMiddleware, push } from 'react-router-redux'

// Create a history of your choosing (we're using a browser history in this case)
const history = createHistory()

// Build the middleware for intercepting and dispatching navigation actions
const middleware = routerMiddleware(history)

// Add the reducer to your store on the `router` key
// Also apply our middleware for navigating
const store = createStore(
  reducers,
  applyMiddleware(middleware)
)

const render = () => {
  ReactDOM.render(
    <AppContainer>
      <Provider store={store}>
      	<ConnectedRouter history={history}>
          <HashRouter>
            <div className={styles.app}>
  	      <Header />
                <div>
                  <Route exact path="/" component={Feed} />
          	  <Route exact path="/add" component={AddStory} />
                </div>
              <Footer />
            </div>
          </HashRouter>
	 </ConnectedRouter>
      </Provider>
    </AppContainer>,
    document.getElementById('root')
  );
};
```

I then at another part of my code attempt to redirect from a container component
```javascript
const mapDispatchToProps = (dispatch, ownProps) => {
  console.log('ownProps', ownProps)

  return {
    submitStory: (story) => {
      dispatch(createStory({
      	title: story.match(/^(.+)\n?/)[1],
      	body: story
      }))
    },
    doAction: (action) => {
      switch(action) {
        case "STORY_ADDED":
          console.log('PUSHING');
          ownProps.history.push('/')
          dispatch(push('/'))
          break;
        default: 
          return null
      }
    }
  }
}
```
There are several issues I'm encountering. One is if I attempt `dispatch(ownProps.history.push('/')` I get an error in the middleware about the `action` being undefined. This is easy to understand since `ownProps.history.push('/')` returns undefined; though it will redirect; but only once. If I try to use the `push` that comes from `react-router-redux` I get a proper `action` object returned, but no redirect with `dispatch(push('/'))`.

As you can see I attempted to remedy the situation by calling `ownProps.history.push('/')` then dispatching with react-router-redux push so I have proper action object but this only works once as well. My component's `shouldComponentUpdate` method never triggers on the state update; even though a new state is clearly produced; adding a new story in the view I'm attempting to redirect to.

Any help would be greatly appreciated.

**Comment by timdorr** [comment]
Read this for the second part: https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/guides/blocked-updates.md

However, this is a bug tracker, not a support system. For usage questions, please use Stack Overflow or Reactiflux. Thanks!
